VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/ImplementEqualsAndGetHashCode/ImplementEqualsAndGetHashCodeCodeFix.cs

```diff
@@ -0,0 +1,256 @@
+
+            if (statement is ClassDeclarationSyntax)
+            {
+                var classDeclaration = (ClassDeclarationSyntax) statement;
+
+                var classSymbol = model.GetDeclaredSymbol(classDeclaration);
+
+                var newRoot = root.ReplaceNode(classDeclaration,
+                    classDeclaration.AddMembers(GetEqualsMethod(model, classDeclaration.Identifier, classDeclaration.Members, classSymbol, objectEquals),
+                        GetGetHashCodeMethod(model, classDeclaration.Members)));
+                return document.WithSyntaxRoot(newRoot);
+            }
+            else
+            {
+                var structDeclaration = (StructDeclarationSyntax)statement;
+
+                var structSymbol = model.GetDeclaredSymbol(structDeclaration);
+
+                var newRoot = root.ReplaceNode(structDeclaration,
+                    structDeclaration.AddMembers(GetEqualsMethod(model, structDeclaration.Identifier, structDeclaration.Members, structSymbol, objectEquals),
+                        GetGetHashCodeMethod(model, structDeclaration.Members)));
+                return document.WithSyntaxRoot(newRoot);
+            }
+        }
+
+        private MethodDeclarationSyntax GetEqualsMethod(SemanticModel model, SyntaxToken identifier, SyntaxList<SyntaxNode> members, INamedTypeSymbol typeSymbol, IMethodSymbol objectEquals)
+        {
+            var publicModifier = SyntaxFactory.Token(SyntaxKind.PublicKeyword);
+            var overrideModifier = SyntaxFactory.Token(SyntaxKind.OverrideKeyword);
+
+            var parameter = SyntaxFactory.Parameter(SyntaxFactory.Identifier("obj"))
+                .WithType(SyntaxFactory.ParseTypeName("object"));
+
+            var ifStatement =
+                SyntaxFactory.IfStatement(SyntaxFactory.ParseExpression($"obj == null || typeof({identifier}) != obj.GetType()"),
+                    SyntaxFactory.Block(SyntaxFactory.ParseStatement("return false;")));
+
+            var castStatement = SyntaxFactory.ParseStatement(
+                    $"var value = ({identifier}) obj;{Environment.NewLine}");
+
+            var fieldAndPropertyEqualityStatements = new List<string>();
+            foreach (var member in members)
+            {
+                if (member.IsKind(SyntaxKind.FieldDeclaration))
+                {
+                    var field = (FieldDeclarationSyntax)member;
+                    if (field.Modifiers.Contains(SyntaxKind.StaticKeyword))
+                    {
+                        continue;
+                    }
+
+                    var symbol = model.GetTypeInfo(field.Declaration.Type).Type;
+                    if (symbol.IsValueType)
+                    {
+                        fieldAndPropertyEqualityStatements.AddRange(field.Declaration.Variables.Select(
+                            variable => $"{variable.Identifier}.Equals(value.{variable.Identifier})"));
```

Comment:
Structs do not implement `==` automatically, leading to a compile error: http://chat.stackexchange.com/transcript/message/29428905#29428905